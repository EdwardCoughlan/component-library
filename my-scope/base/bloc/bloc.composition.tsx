import React from 'react';
import { Bloc, BlocContext } from './bloc';

const ExampleBasicBloc = ({contents}: {contents: string}) =>(
  <div>{contents}</div>
)

export const BasicBloc = () => (
  <Bloc 
  id="basicBloc"
  componentMapperIdentifier="basicBloc"
  props={{unmappedContent: 'hi I am unmapped content' }}
  componentMappers={
    {
      basicBloc: {
        component: ({contents}) => <div>{contents}</div>, 
        mapper: (props) =>({contents: props.unmappedContent.replace('unmapped', 'mapped')})
      }
    }
  }
/>
);


export const ContextBloc = () => (
  <BlocContext.Provider value={{componentMappers: {
    contextBlocComponent: {
      component: ExampleBasicBloc, 
      mapper: ({unmappedContent}) =>{return {contents: `${unmappedContent.replace('unmapped', 'mapped')} from global context`}}
    }
  }}}>
    <Bloc 
      id="basicBloc"
      props={{unmappedContent: 'hi I am unmapped content' }}
      componentMapperIdentifier="contextBlocComponent"
    />
  </BlocContext.Provider>
);
