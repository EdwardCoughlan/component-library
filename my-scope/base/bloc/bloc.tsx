/* eslint-disable @typescript-eslint/no-explicit-any */
import React, {
  createContext,
  createElement,
  FunctionComponentElement,
  useContext,
} from 'react';

export type ComponentMapper = {
  component: React.FC<any>;
  mapper?: (props: any) => any;
};

export type BlocProps = {
  id: string;
  props?: any;
  componentMapperIdentifier: string;
  componentMappers?: Record<string, ComponentMapper>;
};

export const BlocFn = ({
  id,
  props,
  componentMapperIdentifier,
  componentMappers = {},
}: BlocProps):
  | FunctionComponentElement<{
      key: string;
    }>
  | FunctionComponentElement<any> => {
  if (typeof componentMappers[componentMapperIdentifier] === 'undefined') {
    return createElement(
      () => (
        <div>
          The component {componentMapperIdentifier} has not been created yet.
        </div>
      ),
      { key: id }
    );
  }
  const componentMapper = componentMappers[componentMapperIdentifier];
  const componentProps =
    !!props && !!componentMapper.mapper ? componentMapper.mapper(props) : props;
  return createElement(componentMapper.component, {
    ...componentProps,
    key: id,
  });
};

export const BlocContext = createContext<{
  componentMappers?: Record<string, ComponentMapper>;
}>({ componentMappers: undefined });

export function Bloc({
  componentMappers,
  ...restOfProps
}: BlocProps): JSX.Element {
  const context = useContext(BlocContext);
  if (context && context.componentMappers) {
    const contextLevelComponentMappers = context.componentMappers;
    for (const k in componentMappers) {
      const key = k as keyof typeof componentMappers;
      contextLevelComponentMappers[key] = componentMappers[key];
    }
    return BlocFn({
      componentMappers: contextLevelComponentMappers,
      ...restOfProps,
    });
  }
  return BlocFn({ componentMappers, ...restOfProps });
}

