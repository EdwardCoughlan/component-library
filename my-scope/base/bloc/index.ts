export { Bloc, BlocFn, BlocContext } from './bloc';
export type { BlocProps,  ComponentMapper} from './bloc';
