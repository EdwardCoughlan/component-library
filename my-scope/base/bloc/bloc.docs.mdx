---
description: 'A Bloc component.'
labels: ['react', 'typescript', 'base']
---

import { Bloc } from './bloc';

## React Component for rendering text

A basic div that renders some text

### Component usage
```js
<Bloc 
  id="basicBloc"
  componentMapperIdentifier="basicBloc"
  props={{unmappedContent: 'hi I am unmapped content' }}
  componentMappers={
    {
      basicBloc: {
        component: ({contents}) => <div>{contents}</div>, 
        mapper: (props) =>({contents: props.unmappedContent.replace('unmapped', 'mapped')})
      }
    }
  }
/>
```

### Using props to customize the text

Modify the text to see it change live:
```js live
<Bloc 
  id="basicBloc"
  componentMapperIdentifier="basicBloc"
  props={{unmappedContent: 'hi I am unmapped content' }}
  componentMappers={
    {
      basicBloc: {
        component: ({contents}) => <div>{contents}</div>, 
        mapper: (props) =>({contents: props.unmappedContent.replace('unmapped', 'mapped')})
      }
    }
  }
/>
```
