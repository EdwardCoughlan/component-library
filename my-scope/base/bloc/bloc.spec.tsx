import React from 'react';
import { render } from '@testing-library/react';
import { Bloc, BlocContext, ComponentMapper } from './bloc';

const ComponentWithNoProps: React.FC = () => <div>Test with no props</div>;

type ComponentProps = {
  contents: string;
};
type ComponentProps2 = {
  alteredContents: string;
};

const ComponentWithProps: React.FC<ComponentProps> = (props) => (
  <div>{props.contents}</div>
);
const ComponentWithProps2: React.FC<ComponentProps2> = (props) => (
  <div>{props.alteredContents}</div>
);
const ComponentWithProps3: React.FC<ComponentProps2> = (props) => (
  <div>{props.alteredContents}</div>
);

const Mappers: Record<string, (props: unknown) => unknown> = {
  componentWithProps2: (props): ComponentProps2 => {
    return {
      alteredContents: `Altered ${(props as ComponentProps).contents}`,
    };
  },
  componentWithProps3: (): void => {
    throw new Error('Oh shit');
  },
};

const componentMappers: Record<string, ComponentMapper> = {
  componentWithNoProps: { component: ComponentWithNoProps },
  componentWithProps: { component: ComponentWithProps },
  componentWithProps2: {
    component: ComponentWithProps2,
    mapper: Mappers['componentWithProps2'],
  },
  componentWithProps3: {
    component: ComponentWithProps3,
    mapper: Mappers['componentWithProps3'],
  },
};

test('BLoc component without props', () => {
  const { container } = render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentWithNoProps"
      componentMappers={componentMappers}
    />
  );

  expect(container).toMatchSnapshot();
});

test('Bloc component with props', () => {
  const test = "I am a component with props. That doesn't use a mapper";
  const { container } = render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentWithProps"
      props={{
        contents: test,
      }}
      componentMappers={componentMappers}
    />
  );
  expect(container).toMatchSnapshot();
});

test("BLoc component doesn't exists", () => {
  const { container } = render(
    <Bloc
      id="1"
      props={{
        content: "I am a component that doesn't exists",
      }}
      componentMapperIdentifier="componentThatDoesNotExists"
      componentMappers={componentMappers}
    />
  );
  expect(container).toMatchSnapshot();
});

test('Bloc component with props to be mapped', () => {
  const { container } = render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentWithProps2"
      props={{
        contents: 'I am a component with props. That that will be mapped',
      }}
      componentMappers={componentMappers}
    />
  );
  expect(container).toMatchSnapshot();
});

test('Bloc component with props to be mapped error', () => {
  try {
    render(
      <Bloc
        id="1"
        componentMapperIdentifier="componentWithProps3"
        props={{
          contents:
            'I am a component with props. That that will be mapped but an error will happen',
        }}
        componentMappers={componentMappers}
      />
    );
  } catch (err) {
    console.log(err.message);
    expect(err).toBeTruthy();
  }
});

test('Bloc with component coming from BlocContext', () => {
  const ContextComponent = (props: { fromContext: string }) => (
    <div>{`context : ${props.fromContext}`}</div>
  );
  const { container } = render(
    <BlocContext.Provider
      value={{
        componentMappers: {
          componentFromContext: {
            component: ContextComponent,
            mapper: (props) => {
              return {
                fromContext: props.contents,
              };
            },
          },
        },
      }}
    >
      <Bloc
        id="1"
        componentMapperIdentifier="componentFromContext"
        props={{
          contents: 'I am a component that is a part of the global context.',
        }}
        componentMappers={componentMappers}
      />
    </BlocContext.Provider>
  );
  expect(container).toMatchSnapshot();
});

test('No component mappers secified', () => {
  const { container } = render(
    <Bloc
      id="1"
      componentMapperIdentifier="componentFromContext"
      props={{
        contents: 'I am a component that is a part of the global context.',
      }}
    />
  );
  expect(container).toMatchSnapshot();
});
