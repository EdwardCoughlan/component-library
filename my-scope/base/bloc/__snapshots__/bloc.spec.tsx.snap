// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BLoc component doesn't exists 1`] = `
<div>
  <div>
    The component 
    componentThatDoesNotExists
     has not been created yet.
  </div>
</div>
`;

exports[`BLoc component without props 1`] = `
<div>
  <div>
    Test with no props
  </div>
</div>
`;

exports[`Bloc component with props 1`] = `
<div>
  <div>
    I am a component with props. That doesn't use a mapper
  </div>
</div>
`;

exports[`Bloc component with props to be mapped 1`] = `
<div>
  <div>
    Altered I am a component with props. That that will be mapped
  </div>
</div>
`;

exports[`Bloc with component coming from BlocContext 1`] = `
<div>
  <div>
    context : I am a component that is a part of the global context.
  </div>
</div>
`;

exports[`No component mappers secified 1`] = `
<div>
  <div>
    The component 
    componentFromContext
     has not been created yet.
  </div>
</div>
`;
